app/scripts/lib/rpc-method-middleware/handlers/provider-authorize.js

```diff
@@ -0,0 +1,187 @@
+    // What error code and message here?
+    throw new Error(
+      '`requiredScopes` object MUST contain 1 more `scopeObjects`, if present',
+    );
+  }
+
+  const validOptionalScopes = {};
+  for (const [scopeString, scopeObject] of Object.entries(optionalScopes)) {
+    if (isValidScope(scopeString, scopeObject)) {
+      validOptionalScopes[scopeString] = {
+        accounts: [],
+        ...scopeObject,
+      };
+    }
+  }
+  if (optionalScopes && Object.keys(validOptionalScopes).length === 0) {
+    // What error code and message here?
+    throw new Error(
+      '`optionalScopes` object MUST contain 1 more `scopeObjects`, if present',
+    );
+  }
+
+  const randomSessionProperties = {}; // session properties do not have to be honored by the wallet
+  for (const [key, value] of Object.entries(sessionProperties)) {
+    if (Math.random() > 0.5) {
+      randomSessionProperties[key] = value;
+    }
+  }
+  if (sessionProperties && Object.keys(sessionProperties).length === 0) {
+    return end(
+      new EthereumRpcError(5300, 'Invalid Session Properties requested'),
+    );
+  }
+
+  const validScopes = {
+    // what happens if these keys collide?
+    ...validRequiredScopes,
+    ...validOptionalScopes,
+  };
+
+  // TODO:
+  // Unless the dapp is known and trusted, give generic error messages for
+  // - the user denies consent for exposing accounts that match the requested and approved chains,
+  // - the user denies consent for requested methods,
+  // - the user denies all requested or any required scope objects,
+  // - the wallet cannot support all requested or any required scope objects,
+  // - the requested chains are not supported by the wallet, or
+  // - the requested methods are not supported by the wallet
+  // return
+  //     "code": 0,
+  //     "message": "Unknown error"
+
+  if (Object.keys(validScopes).length === 0) {
+    return end(new EthereumRpcError(5000, 'Unknown error with request'));
+  }
+
+  // TODO:
+  // When user disapproves accepting calls with the request methods
+  //   code = 5001
+  //   message = "User disapproved requested methods"
+  // When user disapproves accepting calls with the request notifications
+  //   code = 5002
+  //   message = "User disapproved requested notifications"
+
+  for (const [scopeString] of Object.entries(validScopes)) {
+    if (!isSupportedScopeString(scopeString)) {
+      // A little awkward. What is considered validation? Currently isValidScope only
+      // verifies that the shape of a scopeString and scopeObject is correct, not if it
+      // is supported by MetaMask and not if the scopes themselves (the chainId part) are well formed.
+
+      // Additionally, still need to handle adding chains to the NetworkController and verifying
+      // that a network client exists to handle the chainId
+
+      // Finally, I'm unsure if this is also meant to handle the case where namespaces are not
+      // supported by the wallet.
```

Comment:
yeah we need to figure out where/how providers other than our own own provider can register a list of `supportedMethods`. This would be a hook out to some registry controller which we haven't yet designed for.